src/NamedTypesRegistry.cc

```diff
@@ -22,336 +22,380 @@
+            member = name(el.name.variable);
+        } else if (!el.valueDefinition.typeDefinition.typeSpecification.nestedTypes.empty()) {
+            // - name (array|enum[<nested>, <type>])
+            tn = &el.valueDefinition.typeDefinition.typeSpecification.nestedTypes;
+        } else {
+            ts = &el.sections;
         }
 
-        struct DependencyTypeInfo {
+        if (!member.empty()) {
+            result.insert(member);
+        } else if (ts) {
+            Members sub = collectMembers(*ts);
+            result.insert(sub.begin(), sub.end());
+        } else if (tn) {
+            Members sub = collectMembers(*tn);
+            result.insert(sub.begin(), sub.end());
+        }
+    }
 
-            typedef std::map<std::string, std::string> InheritanceMap;
-            InheritanceMap childToParent;
+    void collectElementMember(const mson::Element::ValueMemberSection& el, Members& result)
+    {
+        std::string member;
+        const mson::TypeSections* ts = NULL;
 
-            typedef std::set<std::string> Members;
-            typedef std::map<std::string, Members> MembersMap;
+        if (!name(el).empty()) {
+            member = name(el);
+        } else {
+            ts = &el.sections;
+        }
 
-            MembersMap objectToMembers;
+        if (!member.empty()) {
+            result.insert(member);
+        } else if (ts) {
+            Members sub = collectMembers(*ts);
+            result.insert(sub.begin(), sub.end());
+        }
+    }
 
-            typedef std::map<std::string, const snowcrash::DataStructure*> ElementMap;
-            ElementMap nameToElement;
+    void collectElementMember(const mson::Element::MixinSection& el, Members& result)
+    {
+        std::string member = name(el);
 
-            const std::string& parent(const snowcrash::DataStructure* ds) const
-            {
-                return ds->typeDefinition.typeSpecification.name.symbol.literal;
-            }
+        if (!member.empty()) {
+            result.insert(std::move(member));
+        }
+    }
 
-            bool hasParent(const snowcrash::DataStructure* ds) const
-            {
-                return !parent(ds).empty();
-            }
+    void collectElementMember(const mson::Element::OneOfSection& el, Members& result) {}
 
-            const std::string& name(const mson::TypeName& tn) const
-            {
-                return tn.symbol.literal;
-            }
+    void collectElementMember(const mson::Element::GroupSection& el, Members& result) {}
 
-            const std::string& name(const snowcrash::DataStructure* ds) const
-            {
-                return ds->name.symbol.literal;
-            }
+    void collectElementMember(const mson::Element::Empty& el, Members& result) {}
 
-            const std::string& name(const mson::TypeSpecification& ts) const
-            {
-                return ts.name.symbol.literal;
-            }
+    struct CollectElementMemberLambda {
```

Comment:
This can be a much shorter lambda in C++14. In C++11 we don't have a choice but introducing a structure with a template operator().